[C++Interop] Allow Multiple SWIFT_CONFORMS_TO_PROTOCOL to C++ Classes

jz709u opened the pull request:
This change introduces the ability for C++ classes to conform to multiple swift protocols using the macro SWIFT_CONFORMS_TO_PROTOCOL.

This resolves https://github.com/swiftlang/swift/issues/75831.

jz709u committed:
[C++Interop] Allow Multiple SWIFT_CONFORMS_TO_PROTOCOL

This change introduces the ability for c++ classes to conform to
multiple swift protocols using the macro SWIFT_CONFORMS_TO_PROTOCOL

egorzhdan reviewed (commented):
Thanks for the pull request @jz709u!
Could you please also add a test case that would verify this behavior? You can find the existing tests for `conforms_to` at `test/Interop/Cxx/class/Inputs/conforms-to.h`, `test/Interop/Cxx/class/conforms-to.swift`, `test/Interop/Cxx/class/conforms-to-errors.swift`.

jz709u committed:
[Test] Multiple Support Tests

- Tests for conforming to protocols for cxx
- Add tests for the derived class DerivedFromHasTestAndHasPlay

egorzhdan reviewed (approved):
This looks awesome, thank you @jz709u!

egorzhdan commented:
@swift-ci please smoke test

jz709u commented:
@egorzhdan I've been on vacation for the past week and haven't been able to work on fixing these issues. I'll examine why the smoke tests fail in two days from today.

egorzhdan commented:
Sounds good, thanks @jz709u!

jz709u commented:
@egorzhdan I merged the main branch to attempt to fix the issues. Since then, the unit test checks have been stuck in yellow. Can you provide any insight as to why the current state is not changing? I appreciate any help you can provide.

egorzhdan commented:
Updating the PR branch resets the unit test status. Let me run the tests again.

Before merging this PR, could you please rebase it on the main branch, instead of having a merge commit? That would help us to have a cleaner history once the PR is merged.

egorzhdan commented:
@swift-ci please smoke test

jz709u commented:
> Updating the PR branch resets the unit test status. Let me run the tests again.
> 
> Before merging this PR, could you please rebase it on the main branch, instead of having a merge commit? That would help us to have a cleaner history once the PR is merged.

@egorzhdan 
Understood, I'll do that now.

jz709u commented:
@egorzhdan Please trigger the smoke test again. I just rebased and squashed the last commit into the test commit.
Thanks Again

egorzhdan commented:
@swift-ci please smoke test

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2937,46 +2937,57 @@ namespace {
-            return false;
-          });
-      if (conformsToAttr == clangDecl->getAttrs().end())
+      SmallVector<clang::Attr*, 40> conformsToMatches;
```
I am not sure what number of conformances do we expect for a class, but 40 sounds a bit too much for me. Note that if we go over that only results in a heap allocation, so this is strictly a performance optimization not a correctness issue if the user has more.

jz709u replied:
Agreed. Do we have data regarding the median amount of conformances?

egorzhdan replied:
I think using `1` as a number of conformances here would be reasonable. There wouldn't usually be many conformances.

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2937,46 +2937,57 @@ namespace {
-                                 .drop_front(StringRef("conforms_to:").size())
-                                 .str();
+      for (auto match : conformsToMatches) {
+        if (auto swiftAttr = dyn_cast<clang::SwiftAttrAttr>(match))
```
Do we ever expect this cast to fail? If we do not, I'd use `cast` instead of `dyn_cast` and remove the if statement.

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2937,46 +2937,57 @@ namespace {
+        return false; 
+      });
+
+      if (conformsToMatches.empty())
```
The for loop below would run 0 times if this vector is empty. We might be able to remove this early return.

Xazax-hun reviewed (commented):
Mostly looks good to me but some nits inline. The tests failed, let's investigate that.

jz709u committed:
Based on feedback remove the if empty check.

- not needed loop will run 0 times

jz709u committed:
Based on feedback remove dynamic cast

jz709u committed:
Move conformances from 40 -> 1 for protocols

Xazax-hun commented:
@swift-ci please smoke test

jz709u committed:
Fix unit test conforms to "error:" matching actual error that should
emit

Xazax-hun commented:
@swift-ci please smoke test

Xazax-hun commented on test/Interop/Cxx/class/conforms-to-errors.swift:
```diff
@@ -15,6 +15,10 @@ struct  __attribute__((swift_attr("conforms_to:Mod.X"))) CModInv {};
 struct  __attribute__((swift_attr("conforms_to:SwiftTest.B"))) CB {};
+struct  __attribute__((swift_attr("conforms_to:SwiftTest.A"))) \
+    __attribute__((swift_attr("conforms_to:SwiftTest.B"))) ConformsToValidAProtocolButAlsoStructB {};
+struct  __attribute__((swift_attr("conforms_to:SwiftTest.A"))) \
```
Sorry for the late comment, I wonder if we want to have a test where the user conforms a struct to the same protocol twice just to make sure there are no crashes or unintended behavior. After that, this patch looks good to me.

jz709u committed:
test duplicate protocol conformance

Xazax-hun commented:
@swift-ci please smoke test

jz709u committed:
Conforming to Non-Duplicate Protocol

jz709u commented:
Please run it once more.

Xazax-hun commented:
@swift-ci please smoke test

jz709u committed:
Duplicate Conformance Protocol Check

jz709u commented:
I have been having issues creating a consistent dev environment so that I can test whether my code runs and works.
This makes development extremely difficult, and I have been working on this for quite a while. The build failures I run into are random, and it is hard to diagnose the cause.

egorzhdan commented:
@swift-ci please smoke test

jz709u commented:
@egorzhdan @Xazax-hun 
What is the expectation here:
this is the error that's being output
> `- error: redundant conformance of 'ConformsToCTwice' to protocol 'C' 

but my test is expecting:
> error: Redundant conformance of 'ConformsToCTwice' to protocol 'SwiftTest.C'

Do we expect the module to be included as a prefix to the protocol name?

jz709u committed:
Include Protocol Module Name in the error for Duplicate Conformance
Error

egorzhdan commented:
I think ideally we would want the module name to be included in the output, e.g. `MySwiftModule.MyProtocol`.

But I don't think it's worth changing that in this patch. Let's preserve the old output format for now, and we can improve it in follow-up patches.

egorzhdan commented:
@swift-ci please smoke test

jz709u committed:
remove e that was accidently added

jz709u committed:
Merge branch 'main' into jayz/allow-multiple-SWIFT_CONFORMS_TO_PROTOCOL

jz709u commented:
@egorzhdan @Xazax-hun  I won't be able to finish this PR.  Thanks for your support.

saehejkang commented:
@egorzhdan can I finish this PR up? What's left to complete?
